src/watcher/chokidar.ts

```diff
@@ -36,25 +40,43 @@ export function watch(
   const watcher = chokidar.watch(paths, options) as FileWatcher
   const programmaticallyWatchedFiles: string[] = []
 
+  const isFileIgnored = (event: string, file: string) => {
+    if (programmaticallyWatchedFiles.includes(file) && isSilentEvent(event)) {
+      log('ignoring file addition because was added silently %s', file)
+      return true
+    }
+
+    log('file watcher event "%s" originating from file/dir %s', event, file)
+    return false
+  }
+
+  // @ts-ignore
+  const originalOnListener = watcher.on
+
   watcher.addSilently = path => {
     programmaticallyWatchedFiles.push(path)
     watcher.add(path)
   }
 
-  if (options && options.onAll) {
-    watcher.on('all', (event, file, stats) => {
-      if (
-        programmaticallyWatchedFiles.includes(file) &&
-        SILENT_EVENTS.includes(event)
-      ) {
-        log('ignoring file addition because was added silently %s', file)
-        return
-      } else {
-        log('file watcher event "%s" originating from file/dir %s', event, file)
-      }
-
-      options.onAll!(event, file, stats)
-    })
+  // Use `function` to bind originalOnListener to the right context
+  watcher.on = function(event: string, listener: (...args: any[]) => void) {
+    if (event === 'all') {
+      return originalOnListener.call(this, event, (eventName, path, stats) => {
+        if (isFileIgnored(eventName, path) === false) {
+          listener(eventName, path, stats)
+        }
+      })
+    }
+
+    if (isSilentEvent(event)) {
```

Comment:
Wracking my brain to understand, is it obvious to you?

Reply:
Refactored to `isSilentableEvent` as suggested below 👍